src/mbgl/text/bidi.cpp

```diff
@@ -0,0 +1,42 @@
+#include <mbgl/text/bidi.hpp>
+#include <mbgl/util/utf.hpp>
+#include <unicode/ushape.h>
+
+namespace mbgl {
+
+BiDi::BiDi()
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+    transform = ubiditransform_open(&errorCode); // Only error is failure to allocate memory, in that case ubidi_transform would fall back to creating transform object on the fly
+}
+    
+BiDi::~BiDi()
+{
+    if ( transform )
+        ubiditransform_close(transform);
+}
+
+std::u32string BiDi::bidiTransform( const std::u32string& input32 )
+{
+    UErrorCode errorCode = U_ZERO_ERROR;
+    
+    std::u16string input16 = util::utf32_to_utf16::convert( input32 );
+    std::unique_ptr<UChar*> outputText = std::make_unique<UChar*>( new UChar[input16.size() * 2] ); // Maximum output of ubidi_transform is twice the size of input according to ubidi_transform.h
```

Comment:
This is what [valgrind is complaining about](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/174564772#L2323) on the Qt5 Travis bot.

I suppose you mean `std::make_unique<UChar[]>(input16.size() * 2);` instead?